### PR TITLE
Automate C++ include file grouping and ordering using clang-format

### LINKED ISSUE #1345: Automate C++ include file grouping and ordering using clang-format
Currently the C++ include files in the RAPIDS codebase are not consistently ordered and grouped.
Inconsistent ordering makes it harder to maintain the codebase, to understand the dependencies,
and to automate changes.

Using clang-format to automate the ordering and grouping of include files will not only add
consistent style, but it will ease automation of changes. For example, We are undertaking a
refactoring of RMM that will replace `rmm::mr::device_memory_resource*` with
`rmm::device_async_resource-ref` everywhere in RAPIDS (not just cuDF). This requires adding an
include to MANY files across multiple RAPIDS repos. Getting the location of this include correct
everywhere is very difficult without automatic grouping of headers.

I propose to use clang-format's `IncludeCategories` settings to automate the ordering and
grouping of include files in the C++ codebase consistently.

Proof-of-concept PRs have been created for RMM and cuDF:
 - cuDF: https://github.com/rapidsai/cudf/pull/15063
 - RMM: https://github.com/rapidsai/rmm/pull/1463

The ordering used for cuDF is given by the following settings.

```yaml
IncludeBlocks: Regroup
IncludeCategories:
  - Regex:           '^"' # quoted includes
    Priority:        1
  - Regex:           '^<(benchmarks|tests)/' # benchmark includes
    Priority:        2
  - Regex:           '^<cudf_test/' # cuDF includes
    Priority:        3
  - Regex:           '^<cudf/' # cuDF includes
    Priority:        4
  - Regex:           '^<(nvtext|cudf_kafka)' # other libcudf includes
    Priority:        5
  - Regex:           '^<(cugraph|cuml|cuspatial|raft|kvikio)' # Other RAPIDS includes
    Priority:        6
  - Regex:           '^<rmm/' # RMM includes
    Priority:        7
  - Regex:           '^<(thrust|cub|cuda)/' # CCCL includes
    Priority:        8
  - Regex:           '^<(cooperative_groups|cuco|cuda.h|cuda_runtime|device_types|math_constants|nvtx3)' # CUDA includes
    Priority:        8
  - Regex:           '^<.*\..*' # other system includes (e.g. with a '.')
    Priority:        9
  - Regex:           '^<[^.]+' # STL includes (no '.')
    Priority:        10
```

### PR DESCRIPTION
**Pull request by harrism**
## Description
This uses the IncludeCategories settings in .clang-format to attempt to enforce our documented #include order in libcuspatial. For discussion, see https://github.com/rapidsai/cudf/pull/15063. This PR uses a subset of the header grouping categories used in that PR.

Note that this also updates the pre-commit configuration to check and correct file copyright years. This comes from https://github.com/rapidsai/cudf/pull/14917.  @KyleFromNVIDIA can you confirm whether or not it's OK to do this? It seems to be working fine.

Closes #1345 

## Checklist
- [X] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cuspatial/blob/HEAD/CONTRIBUTING.md).
- [X] New or existing tests cover these changes.
- [X] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by KyleFromNVIDIA** [comment]
+1 to the new copyright update script, we're planning to eventually roll that out to all RAPIDS repos.

File: cpp/.clang-format [hunk 1]
```diff
@@ -71,8 +71,28 @@ ForEachMacros:
...
+IncludeCategories:
+  - Regex:           '^"' # quoted includes
+    Priority:        1
+  - Regex:           '^<(benchmarks|tests)/' # benchmark includes
+    Priority:        2
+  - Regex:           '^<(cuspatial_test|cuproj_test)/' # cuSpatial/cuProj test includes
+    Priority:        3
+  - Regex:           '^<(cuspatial|cuproj)/' # cuSpatial/cuProj includes
+    Priority:        4
+  - Regex:           '^<(cudf_test|cudf|cuml|raft|kvikio)' # Other RAPIDS includes
```
Comment (isVoid) [inline cpp/.clang-format:84]: I thought about removing cuml/raft/kvikio since they are not used, should it?

**Review by isVoid — APPROVED** [review]
One small question below.

File: cpp/.clang-format (hunk 1 above)
Comment (harrism) [reply cpp/.clang-format:84]: We could, but I'm also trying to keep things relatively consistent across RAPIDS repos. But I'm not confident in that direction. Thoughts?

Comment (isVoid) [reply cpp/.clang-format:84]: Yeah I'm not confident either - especially if we apply this remove-not-used rule, then what about the cuda libraries below? If we remove cuda rules, are we sure that future contributors do not use them? 

From that end, I think the future-forward approach is to maintain the disjunction set from all RAPIDS libraries to embed the collective behavioral customs. This reduces possible conflicts when migrating code or merging repos.

**Pull request merged**